### ISSUE TITLE
Multiple air data API calls during a single polling interval

### ISSUE THREAD

**Issue by coughski** [DMBlakeley/homebridge-awair2#66]
Hi,

Thanks for providing such a useful Homebridge plugin. I think I've encountered an issue here which results in stale sensor data being displayed for hours. Please let me know if you need any additional information.

**Describe The Bug:**
I'm running an older version of Homebridge (v1.1.7) on an older Mac (v10.11.6). 
According to the Homebridge logs, it looks like there are 4 API calls being made per polling_interval, instead of just one. You can see there are 4 practically identical updateAirData json responses within a second of each other, at 7:03:46 AM and 7:03:47 AM. 
I think these duplicate API calls are leading to 429 status codes a short time later at 7:18:46 AM, since my API quota limits are probably being reached with these duplicate calls.

**Expected behavior:**
A single API call per polling_interval so that the Hobbyist quota limits are not exceeded, which should prevent status code 429 from appearing, and allow sensor data in Homebridge to be properly refreshed, instead of displaying stale values.

**Logs:**

```
[2/19/2021, 6:48:52 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] updateAirData: [{"timestamp":"2021-02-19T12:00:00.000Z","score":87.63636363636364,"sensors":[{"comp":"pm25","value":0},{"comp":"temp","value":20.025454781272195},{"comp":"humid","value":25.766363577409223},{"comp":"voc","value":294},{"comp":"co2","value":443.59090909090907}],"indices":[{"comp":"co2","value":0},{"comp":"pm25","value":0},{"comp":"voc","value":0},{"comp":"temp","value":0},{"comp":"humid","value":-2}]}]
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "temp": 20.025454781272195
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "humid": 25.766363577409223
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "co2": 443.59090909090907
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] aqi array: [1,0,0,1,0]
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] ppb => ug/m^3 equation: (294 * 72.6657827301974 * 1 * 101.32) / ((273.15 + 20.025454781272195) * 8.3144)
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112]: voc (294 ppb) => tvoc (888.0020036833567 ug/m^3)
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] CO2Before: 0
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] CO2 NORMAL: 443.59090909090907 < 800
[2/19/2021, 7:03:46 AM] [Awair2] [70886B123112] Carbon Dioxide already low.
[2/19/2021, 7:03:47 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] updateAirData: [{"timestamp":"2021-02-19T12:00:00.000Z","score":87.63636363636364,"sensors":[{"comp":"pm25","value":0},{"comp":"humid","value":25.766363577409223},{"comp":"temp","value":20.025454781272195},{"comp":"voc","value":294},{"comp":"co2","value":443.59090909090907}],"indices":[{"comp":"temp","value":0},{"comp":"pm25","value":0},{"comp":"co2","value":0},{"comp":"voc","value":0},{"comp":"humid","value":-2}]}]
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] ppb => ug/m^3 equation: (294 * 72.6657827301974 * 1 * 101.32) / ((273.15 + 20.025454781272195) * 8.3144)
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112]: voc (294 ppb) => tvoc (888.0020036833567 ug/m^3)
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] updateAirData: [{"timestamp":"2021-02-19T12:00:00.000Z","score":87.63636363636364,"sensors":[{"comp":"pm25","value":0},{"comp":"humid","value":25.766363577409223},{"comp":"voc","value":294},{"comp":"co2","value":443.59090909090907},{"comp":"temp","value":20.025454781272195}],"indices":[{"comp":"pm25","value":0},{"comp":"voc","value":0},{"comp":"humid","value":-2},{"comp":"co2","value":0},{"comp":"temp","value":0}]}]
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] ppb => ug/m^3 equation: (294 * 72.6657827301974 * 1 * 101.32) / ((273.15 + 20.025454781272195) * 8.3144)
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112]: voc (294 ppb) => tvoc (888.0020036833567 ug/m^3)
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] updateAirData: [{"timestamp":"2021-02-19T12:00:00.000Z","score":87.63636363636364,"sensors":[{"comp":"pm25","value":0},{"comp":"humid","value":25.766363577409223},{"comp":"temp","value":20.025454781272195},{"comp":"voc","value":294},{"comp":"co2","value":443.59090909090907}],"indices":[{"comp":"voc","value":0},{"comp":"temp","value":0},{"comp":"pm25","value":0},{"comp":"humid","value":-2},{"comp":"co2","value":0}]}]
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "humid": 25.766363577409223
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "temp": 20.025454781272195
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "co2": 443.59090909090907
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] aqi array: [1,0,0,1,0]
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] ppb => ug/m^3 equation: (294 * 72.6657827301974 * 1 * 101.32) / ((273.15 + 20.025454781272195) * 8.3144)
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112]: voc (294 ppb) => tvoc (888.0020036833567 ug/m^3)
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] CO2Before: 0
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] CO2 NORMAL: 443.59090909090907 < 800
[2/19/2021, 7:03:47 AM] [Awair2] [70886B123112] Carbon Dioxide already low.
[2/19/2021, 7:03:48 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:03:49 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:03:50 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:03:51 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:03:52 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:03:54 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:03:58 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:18:46 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] updateAirData: [{"timestamp":"2021-02-19T12:15:00.000Z","score":87.65217391304348,"sensors":[{"comp":"co2","value":435.5217391304348},{"comp":"pm25","value":0},{"comp":"voc","value":288.3478260869565},{"comp":"humid","value":25.71173900106679},{"comp":"temp","value":19.95217389645784}],"indices":[{"comp":"voc","value":0},{"comp":"humid","value":-2},{"comp":"pm25","value":0},{"comp":"temp","value":0},{"comp":"co2","value":0}]}]
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "co2": 435.5217391304348
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "humid": 25.71173900106679
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "temp": 19.95217389645784
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] aqi array: [0,1,1,0,0]
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] CO2Before: 0
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] CO2 NORMAL: 435.5217391304348 < 800
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] Carbon Dioxide already low.
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] ppb => ug/m^3 equation: (288.3478260869565 * 72.6657827301974 * 1 * 101.32) / ((273.15 + 19.95217389645784) * 8.3144)
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112]: voc (288.3478260869565 ppb) => tvoc (871.1478413298362 ug/m^3)
[2/19/2021, 7:18:46 AM] [Awair2] updateAirData error: Error: Request failed with status code 429
[2/19/2021, 7:18:46 AM] [Awair2] updateAirData error: Error: Request failed with status code 429
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] updateAirData: [{"timestamp":"2021-02-19T12:15:00.000Z","score":87.65217391304348,"sensors":[{"comp":"temp","value":19.95217389645784},{"comp":"voc","value":288.3478260869565},{"comp":"pm25","value":0},{"comp":"humid","value":25.71173900106679},{"comp":"co2","value":435.5217391304348}],"indices":[{"comp":"voc","value":0},{"comp":"temp","value":0},{"comp":"humid","value":-2},{"comp":"co2","value":0},{"comp":"pm25","value":0}]}]
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "temp": 19.95217389645784
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "humid": 25.71173900106679
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] convertAwairAqi ignoring "co2": 435.5217391304348
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] aqi array: [0,1,1,0,0]
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] ppb => ug/m^3 equation: (288.3478260869565 * 72.6657827301974 * 1 * 101.32) / ((273.15 + 19.95217389645784) * 8.3144)
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112]: voc (288.3478260869565 ppb) => tvoc (871.1478413298362 ug/m^3)
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] CO2Before: 0
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] CO2 NORMAL: 435.5217391304348 < 800
[2/19/2021, 7:18:46 AM] [Awair2] [70886B123112] Carbon Dioxide already low.
[2/19/2021, 7:18:47 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:18:47 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:18:48 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:18:49 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:18:50 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:18:51 AM] [Awair2] [70886B123112] batteryLevel: 99 batteryPlugged: true lowBattery: false
[2/19/2021, 7:18:53 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:33:45 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:33:45 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:33:45 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:33:45 AM] [Awair2] [70886B123112] Updating status...awair-omni_12887
[2/19/2021, 7:33:47 AM] [Awair2] LocalAPI lux data for awair-omni: 0
[2/19/2021, 7:33:47 AM] [Awair2] updateAirData error: Error: Request failed with status code 429
[2/19/2021, 7:33:47 AM] [Awair2] updateAirData error: Error: Request failed with status code 429
[2/19/2021, 7:33:47 AM] [Awair2] updateAirData error: Error: Request failed with status code 429
[2/19/2021, 7:33:47 AM] [Awair2] updateAirData error: Error: Request failed with status code 429
```

**Plugin Config:**

```json
{
    "bridge": {
        "name": "Homebridge",
        "username": "redacted",
        "port": 51826,
        "pin": "redacted"
    },
    "description": "Homebridge config",
    "accessories": [],
    "platforms": [
        {
            "name": "Config",
            "port": 8581,
            "auth": "form",
            "theme": "auto",
            "tempUnits": "f",
            "lang": "auto",
            "platform": "config"
        },
        {
            "token": "redacted",
            "userType": "users/self",
            "airQualityMethod": "awair-aqi",
            "endpoint": "15-min-avg",
            "limit": 1,
            "carbonDioxideThreshold": 1000,
            "carbonDioxideThresholdOff": 800,
            "occupancyDetection": false,
            "occupancyOffset": 2,
            "occupancyRestart": false,
            "logging": true,
            "verbose": true,
            "enableModes": true,
            "development": false,
            "platform": "Awair2"
        }
    ]
}
```

**Environment:**

* **Plugin Version**: v5.8.1
* **Homebridge Version**: v1.1.7
* **Node.js Version**: v14.15.5
* **NPM Version**: v7.5.4
* **Operating System**: v10.11.6 (OS X El Capitan)

**Comment by deanlyoung** [comment]
Hi @coughski what device_type (i.e. model) do you have?

**Comment by coughski** [comment]
Hi @deanlyoung it's an Awair Omni

**Comment by DMBlakeley** [comment]
Hi @coughski, I do not believe that your environment versions are causing the problem. Agree that the error 429 is due to the 4 calls exceeding the Hobbyist limits. Looking at my logs I also see that I am getting 4 calls for each device but as I have a Developer account I unfortunately missed that there may be an issue in the code. Let me review and will get back to you.

**Comment by coughski** [comment]
Thanks @DMBlakeley much appreciated! 🙏🏻

**Comment by DMBlakeley** [comment]
Hi @coughski, I have released v5.8.2 which should address the issue that you found. Problem was due to accessories being duplicated causing the multiple calls which you observed. 

Please give v5.8.2 a go and let me know if your issue is addressed. If you have left the plugin running it may take some time for the error 429 to stop before things return to normal.

**Comment by DMBlakeley** [comment]
Closing issue. Confirmed that v5.8.2 performs single API call per polling_interval so that the Hobbyist quota limits are not exceeded on MacOS and RaspPi configurations.

Thank you for submitting the problem report and patience while being addressed.

**Comment by coughski** [comment]
Thanks @DMBlakeley for the quick turnaround! I can confirm 15-min and 5-min averages are working as expected (1 updateAirData call per 15/5 mins respectively), but the behavior for latest and raw confuses me. For both of these modes, I see one updateAirData call per minute, instead of per every 5 minutes and 3.3 minutes respectively. Did I misunderstand how those 2 modes work? Calling once a minute in those modes would also exceed Hobbyist quotas.

**Comment by deanlyoung** [comment]
@coughski FYI - I also updated your tier as a thank you for reporting the issue, and also to hopefully give you a buffer on number of calls.

**Comment by coughski** [comment]
Thank you!

**Comment by DMBlakeley** [comment]
@coughski, much thanks for being so thorough in your review of the plugin!

I have reviewed the code and it is correct but assumes that `userTier` is `Hobbyist` in the `Get User Info` call per the API documentation. Appears that this is not the case and why you are seeing the shorter interval than expected.

@deanlyoung, can you check the response for `Hobbyist` tier?

**Comment by deanlyoung** [comment]
@DMBlakeley regarding the /raw endpoint frequency, the way I used these in homebridge-awair was to be able to define your own averaging. For example, a "limit" of ~6 × 10-second samples would average for a 1-minute average, and then the polling frequency was independent of that, but would need to take into account the API Tier quotas of the account. So it might not evenly divide up the day into continuous, discreet blocks, but you could have more recent data at each polling. In other words, a 1-minute average pulled every 5 minutes might pass through spikes as opposed to 5-minute or 15-minute averages at the same frequency. My personal favorite is a 2-minute average. Either that or much more frequent polling via the Local API to save on external bandwidth, but have high availability / near-instantaneous updates. A lot of possibilities, for sure...

**Comment by deanlyoung** [comment]
Also, I set @coughski's Tier to Small Developer, but I don't remember the quota for each air-data endpoint off the top of my head...

**Comment by DMBlakeley** [comment]
The current code was intended to make sure that API User Tier limits were not violated. Problem is that 'Hobbyist' is not being returned so the logic fails.

**Comment by deanlyoung** [comment]
Ah, so in trying to alleviate the issue, I made another... 🤦🏻‍♂️
Adding the other Tiers might be good (HOBBYIST, SMALL_DEVELOPER, LARGE_DEVELOPER, ENTERPRISE) as well as a fallback just in case (maybe to hobbyist?)

**Comment by DMBlakeley** [comment]
I do read the tier limits from the user account so protected there. Problem only with Hobbyist, see email which I sent you with details.

**Comment by DMBlakeley** [comment]
@coughski, I tried to reproduce the behavior that you reported for 'raw' and 'latest' endpoints. @deanlyoung provided me with a test Hobbyist token to so that I could test the code and I found the behavior to be as expected.

I believe what happened was @deanlyoung updated your account from Hobbyist to Small Developer after the initial report that I corrected in v5.8.2 but before you did the 'raw' and 'latest' checks. The behavior you saw was a result of the higher limits with this account setting. As you are now set with this higher tier you should not experience any similar issues.

I am going to close this issue for now. Appreciate you letting me know any additional issues that you find.

Regards, Doug

**Comment by coughski** [comment]
@deanlyoung @DMBlakeley thanks for all your help!
